internal/operands/common-instancetypes/reconcile.go

```diff
@@ -21,90 +30,270 @@ const (
+	resourceFromURLByName := make(map[string]instancetypev1alpha2.VirtualMachineClusterPreference)
+	for _, resourceFromURL := range resourcesFromURL {
+		resourceFromURLByName[resourceFromURL.Name] = resourceFromURL
 	}
+	for _, existingResource := range existingResources {
+		if _, resourceProvided := resourceFromURLByName[existingResource.Name]; !resourceProvided {
+			request.Logger.Info(fmt.Sprintf("removing the no longer provided %s VirtualMachineClusterPreference", existingResource.Name))
+			if err := request.Client.Delete(request.Context, &existingResource); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
 }
 
-func (c *commonInstancetypes) Reconcile(request *common.Request) ([]common.ReconcileResult, error) {
-	request.Logger.Info("Reconciling common-instancetypes")
+func (c *CommonInstancetypes) reconcileFromURL(request *common.Request) ([]common.ReconcileResult, error) {
+	// TODO - In the future we should handle cases where the URL remains the same but the provided resources change.
+	if c.resourceURL != "" && c.resourceURL == *request.Instance.Spec.CommonInstancetypes.URL {
+		request.Logger.Info(fmt.Sprintf("Skipping reconcile of common-instancetypes from URL %s, force with a restart of the service.", *request.Instance.Spec.CommonInstancetypes.URL))
+		return nil, nil
+	}
+
+	c.resourceURL = *request.Instance.Spec.CommonInstancetypes.URL
+	request.Logger.Info(fmt.Sprintf("Reconciling common-instancetypes from URL %s", c.resourceURL))
+	clusterInstancetypesFromURL, clusterPreferencesFromURL, err := c.FetchResourcesFromURL(c.resourceURL)
+	if err != nil {
+		return nil, err
+	}
+
+	existingClusterInstancetypes, existingClusterPreferences, err := c.fetchExistingResources(request)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = reconcileRemovedInstancetypes(request, existingClusterInstancetypes, clusterInstancetypesFromURL); err != nil {
+		return nil, err
+	}
+
+	if err = reconcileRemovedPreferences(request, existingClusterPreferences, clusterPreferencesFromURL); err != nil {
+		return nil, err
+	}
+
+	// Generate the normal set of reconcile funcs to create or update the provided resources
+	c.virtualMachineClusterInstancetypes = clusterInstancetypesFromURL
+	c.virtualMachineClusterPreferences = clusterPreferencesFromURL
+	return common.CollectResourceStatus(request, c.reconcileFuncs()...)
+}
+
+func (c *CommonInstancetypes) reconcileFromBundle(request *common.Request) ([]common.ReconcileResult, error) {
+	request.Logger.Info("Reconciling common-instancetypes from internal bundle")
+	clusterInstancetypesFromBundle, clusterPreferencesFromBundle, err := c.fetchResourcesFromBundle()
+	if err != nil {
+		return nil, err
+	}
+
+	existingClusterInstancetypes, existingClusterPreferences, err := c.fetchExistingResources(request)
```

Comment:
Move removal of existing resources into separate func to avoid duplication? Or what do you think of moving L271-L286 into `Reconcile` func and returning the new resources from `reconcileFromURL` and `reconcileFromBundle`?

Reply:
I'd rather have a shared call in `reconcileFromBundle` and `reconcileFromURL` tbh, so we remove first and then create/update.